staging/src/k8s.io/apiextensions-apiserver/test/integration/ratcheting_test.go

```diff
@@ -0,0 +1,1345 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"testing"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apiextensions-apiserver/pkg/features"
+	"k8s.io/apiextensions-apiserver/test/integration/fixtures"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/dynamic"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+)
+
+var stringSchema *apiextensionsv1.JSONSchemaProps = &apiextensionsv1.JSONSchemaProps{
+	Type: "string",
+}
+
+var stringMapSchema *apiextensionsv1.JSONSchemaProps = &apiextensionsv1.JSONSchemaProps{
+	Type: "object",
+	AdditionalProperties: &apiextensionsv1.JSONSchemaPropsOrBool{
+		Schema: stringSchema,
+	},
+}
+
+var numberSchema *apiextensionsv1.JSONSchemaProps = &apiextensionsv1.JSONSchemaProps{
+	Type: "integer",
+}
+
+var numbersMapSchema *apiextensionsv1.JSONSchemaProps = &apiextensionsv1.JSONSchemaProps{
+	Type: "object",
+	AdditionalProperties: &apiextensionsv1.JSONSchemaPropsOrBool{
+		Schema: numberSchema,
+	},
+}
+
+type RatchetingTestContext struct {
+	*testing.T
+	DynamicClient       dynamic.Interface
+	APIExtensionsClient clientset.Interface
+}
+
+type RatchetingTestOperation = func(ctx *RatchetingTestContext) error
+
+func ExpectError(op RatchetingTestOperation) RatchetingTestOperation {
+	return func(ctx *RatchetingTestContext) error {
+		err := op(ctx)
+		if err != nil {
+			return nil
+		}
+		return errors.New("expected error")
+	}
+}
+
+// apiextensions-apiserver has discovery disabled, so hardcode this mapping
+var fakeRESTMapper map[schema.GroupVersionResource]string = map[schema.GroupVersionResource]string{
+	noxuV1Beta1: "WishIHadChosenNoxu",
+}
+
+// Applies a patch to given GVR and waits for it to be reflected by a GET
+func ApplyPatchOperation(gvr schema.GroupVersionResource, name string, patch map[string]interface{}) RatchetingTestOperation {
```

Comment:
lower-case

Reply:
and everywhere else

Reply:
removed all exported symbols from the test and changed the package to integration_test